### PR TITLE
SLVS-851 S3264 - Unused "events" should be removed

### PR DESCRIPTION
**Pull request by tvsonar**

### CODE REVIEW

File: src/SonarLint.CSharp/Rules/UninvokedPrivateEventDeclaration.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                        }
+
+                        if (!invokedSymbols.Contains(removableEvent.Symbol))
+                        {
+                            var eventField = removableEvent.SyntaxNode as VariableDeclaratorSyntax;
+                            var location = eventField != null
+                                ? eventField.Identifier.GetLocation()
+                                : ((EventDeclarationSyntax)removableEvent.SyntaxNode).Identifier.GetLocation();
+
+                            c.ReportDiagnostic(Diagnostic.Create(Rule, location));
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/UninvokedPrivateEventDeclaration.cs]: ReportDiagnosticIfNonGenerated

File: src/SonarLint.CSharp/Rules/UninvokedPrivateEventDeclaration.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+namespace SonarLint.Rules.CSharp
+{
+    using ClassWithSemanticModel = SyntaxNodeSemanticModelTuple<ClassDeclarationSyntax>;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [SqaleConstantRemediation("10min")]
+    [SqaleSubCharacteristic(SqaleSubCharacteristic.Understandability)]
+    [Rule(DiagnosticId, RuleSeverity, Title, IsActivatedByDefault)]
+    [Tags(Tag.Unused)]
+    public class UninvokedPrivateEventDeclaration : SonarDiagnosticAnalyzer
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/UninvokedPrivateEventDeclaration.cs]: This should also work on public events.

File: src/SonarLint.CSharp/Rules/UninvokedPrivateEventDeclaration.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+                    foreach (var removableEvent in allRemovableEvents)
+                    {
+                        if (!usedSymbols.Contains(removableEvent.Symbol))
+                        {
+                            /// reported by <see cref="UnusedPrivateMember"/>
+                            continue;
+                        }
+
+                        if (!invokedSymbols.Contains(removableEvent.Symbol))
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/UninvokedPrivateEventDeclaration.cs]: We need to be a bit more conservative: should check if the event was assigned to anything (assigned, argument, assigned in field initializer)

File: src/SonarLint.CSharp/Rules/UnusedPrivateMember.cs [hunk 4]
```diff
@@ -32,6 +32,9 @@
 
 namespace SonarLint.Rules.CSharp
 {
+    using ClassWithSemanticModel = SyntaxNodeSemanticModelTuple<ClassDeclarationSyntax>;
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/UnusedPrivateMember.cs]: these can be removed, they are not used any more.

**Comment by dbolkensteyn** [comment]
ok i trust you on this one - the PR is a bit large to review

**Pull request merged**